NumLib/NewtonMethods/NewtonRaphson.h

```diff
@@ -0,0 +1,79 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef NUMLIB_NEWTONMETHODS_NEWTONRAPHSON_H_
+#define NUMLIB_NEWTONMETHODS_NEWTONRAPHSON_H_
+
+#include <logog/include/logog.hpp>
+
+#include <Eigen/Dense>
+
+namespace NumLib
+{
+
+/// Newton-Raphson solver for system of equations using an Eigen linear solvers
+/// library.
+/// The current implementation does not update the solution itself, but calls a
+/// function for the solution's update with the current increment.
+template <typename LinearSolver, typename JacobianMatrix,
+          typename JacobianMatrixUpdate, typename ResidualVector,
+          typename ResidualUpdate, typename SolutionUpdate>
+class Newton final
+{
+public:
+    Newton(LinearSolver& linear_solver, JacobianMatrixUpdate jacobian_update,
+           ResidualUpdate residual_update, SolutionUpdate solution_update,
+           int const maximum_iterations, double const tolerance_squared)
+        : _linear_solver(linear_solver),
+          _jacobian_update(jacobian_update),
+          _residual_update(residual_update),
+          _solution_update(solution_update),
+          _maximum_iterations(maximum_iterations),
+          _tolerance_squared(tolerance_squared)
+    {}
+
+    /// Returns number of iterations.
+    int solve(JacobianMatrix& jacobian, ResidualVector& residual) const
```

Comment:
Hmm, this is implicitly handled by the `*_update` functions. The reason behind it is, that the "solution" (at least in Burgers) is spread among several quantities: see MaterialLib/SolidModels/ViscoElasticBurgers-impl.h in the #1391 PR around line 85 (update_solution).
Better to discuss further things...